src/data/equiv/derangements/finite.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.derangements.basic
+import data.fintype.basic
+import data.fintype.card
+import data.set.finite
+
+import tactic.ring
+import tactic.zify
+
+/-!
+# Derangements on fintypes
+
+This file contains lemmas that describe the cardinality of `derangements α` when `α` is a fintype.
+
+# Main definitions
+
+  - `num_derangements n`: The number of derangements on an n-element set. For concreteness, `fin n`
+    is used.
+
+  - `num_derangements_invariant`: A lemma stating that the number of derangements on a type `α`
+    depends only on the cardinality of `α`.
+
+  - `num_derangements_recursive`: A lemma giving a recursive expression for `num_derangements (n+2)`
+    in terms of `num_derangements n` and `num_derangements (n+1)`.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
+-/
+
+open equiv fintype perm
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (@is_derangement α) := λ _, fintype.decidable_forall_fintype
+
+instance : fintype (derangements α) :=
+begin
+  rw [derangements],
+  apply_instance
+end
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
+
+@[simp] lemma num_derangements_invariant (α : Type*) [fintype α] [decidable_eq α] :
+  card (derangements α) = num_derangements (card α) :=
+begin
+  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
+  use derangements_congr (equiv_fin α),
+end
+
+theorem num_derangements_succ (n : ℕ) :
```

Comment:
I think it would be nice to use this inductive relationship as a definition and then to show `num_derangements_invariant` from there. That shouldn't change too much the proofs, it gets you rid of this weird reference to `fin n`, and allows for better evaluation (I think?).

Reply:
After reflexion, a better inductive definition might be
```
def num_derangements : ℕ → ℕ
| 0       := 1
| (n + 1) := (n + 1) * num_derangements n + (-1 : ℤ)^n
```
(it doesn't work as of because -1 isn't a natural, but that shouldn't be too hard to overcome, maybe by splitting cases on parity)

Reply:
It's rather tricky to go about that -- the reason the proof for `num_derangements_succ` goes through is that we can ignore the complicated specifics of the type (`fin(n+1)` minus a particular element), and instead just consider the cardinality.

I tried something with `induction h_card : (card α) generalizing α`, and it's a lot less clear. Maybe I'll just replace `num_derangements` with `num_derangements_aux` and then show that your definition matches the original.